src/test/kotlin/org/eln2/libelectric/sim/thermal/ThermalTests.kt

```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Test
+    fun consistent_data_model() {
+        val sim = Simulator(TestEnv(0.0))
+        val a = TestBody(Loc(0), ThermalMass(Material.IRON))
+        val b = TestBody(Loc(0), ThermalMass(Material.IRON))
+        val c = TestBody(Loc(0), ThermalMass(Material.IRON))
+        assert(a !in sim.bodies)
+        assert(b !in sim.bodies)
+        assert(c !in sim.bodies)
+        val conn = sim.connect(a, b)
+        assert(conn in sim.connections)
+        assertEquals(conn.a, a)
+        assertEquals(conn.b, b)
+        assert(a in sim.bodies)
+        assert(b in sim.bodies)
+        assert(c !in sim.bodies)
+        sim.remove(conn)
+        assert(conn !in sim.connections)
+        assert(a in sim.bodies)
+        assert(b in sim.bodies)
+        assert(c !in sim.bodies)
+        sim.remove(a, b)
+        assert(a !in sim.bodies)
+        assert(b !in sim.bodies)
+        assert(c !in sim.bodies)
+        val conn2 = sim.connect(b, c)
+        assert(conn2 in sim.connections)
+        assert(a !in sim.bodies)
+        assert(b in sim.bodies)
+        assert(c in sim.bodies)
+        sim.remove(b)
+        assert(conn2 !in sim.connections)
+        assert(a !in sim.bodies)
+        assert(b !in sim.bodies)
+        assert(c in sim.bodies)
+    }
+
+    @Test
+    fun perfectly_isolated() {
+        val sim = Simulator(TestEnv(0.0))
+        val a = TestBody(Loc(1), ThermalMass(Material.IRON))
+        val b = TestBody(Loc(2), ThermalMass(Material.IRON))
+        a.mass.temperature = STANDARD_TEMPERATURE
+        b.mass.temperature = STANDARD_TEMPERATURE + A_BIT
+        sim.add(a, b)
+        val a_old = a.mass.temperature
+        val b_old = b.mass.temperature
+        sim.step(0.05)
+        assert(a in sim.bodies)
+        assert(b in sim.bodies)
+        assertEquals(a.mass.temperature, a_old)
+        assertEquals(b.mass.temperature, b_old)
+    }
+
+    @Test
+    fun second_law() {
+        val sim = Simulator(TestEnv(0.0))  // Don't conduct from/to env
+        val a = TestBody(Loc(1), ThermalMass(Material.IRON))
+        val b = TestBody(Loc(2), ThermalMass(Material.IRON))
+        val lesser = STANDARD_TEMPERATURE
+        val greater = STANDARD_TEMPERATURE + A_BIT
+        for((a_temp, b_temp) in listOf(lesser to greater, greater to lesser)) {
+            a.mass.temperature = a_temp
+            b.mass.temperature = b_temp
+            val a_old = a.mass.temperature
+            val b_old = b.mass.temperature
+            val conn = sim.connect(a, b)
+            sim.step(0.05)
+            assertEquals(a_temp.compareTo(b_temp), a_old.compareTo(a.mass.temperature)) {"A old $a_old, now ${a.mass.temperature}" }
+            assertEquals(b_temp.compareTo(a_temp), b_old.compareTo(b.mass.temperature)) {"B old $b_old, now ${b.mass.temperature}" }
+        }
+    }
+
+    @Test
+    fun conductivity_proportional_to_connectivity() {
+        var last_increase: Temperature? = null
+        for(it in 1..10) {
```

Comment:
kotlinize.

Reply:
This one was _specifically_ requested by Codacy, which claims that the performance of `(1..10).forEach` is strictly worse than this, and lints it